docs/site/Middleware.md

```diff
@@ -0,0 +1,335 @@
+  protected invokeMiddleware: InvokeMiddleware = () => false;
+
+  // ...
+  async handle(context: RequestContext): Promise<void> {
+    try {
+      const {request, response} = context;
+      const finished = await this.invokeMiddleware(context);
+      if (finished) return;
+      const route = this.findRoute(request);
+      const args = await this.parseParams(request, route);
+      const result = await this.invoke(route, args);
+
+      debug('%s result -', route.describe(), result);
+      this.send(response, result);
+    } catch (error) {
+      this.reject(context, error);
+    }
+  }
+}
+```
+
+#### Extend sequence with more than one middleware actions
+
+Sometimes we want to add middleware to the sequence in between other actions,
+for example, do some post-processing before the result is written to the HTTP
+response. This can be achieved by overriding the sequence implementation to call
+`invokeMiddleware` as needed.
+
+{% include code-caption.html content="src/sequence.ts" %}
+
+```ts
+export class SequenceWithMiddleware extends DefaultSequence {
+  async handle(context: RequestContext): Promise<void> {
+    try {
+      const {request, response} = context;
+      // The default middleware chain
+      let finished = await this.invokeMiddleware(context);
+      if (finished) return;
+      const route = this.findRoute(request);
+      const args = await this.parseParams(request, route);
+      const result = await this.invoke(route, args);
+
+      // The second middleware chain for post-invocation processing
+      context.bind('invocation.result').to(result);
+      // POST_INVOCATION_MIDDLEWARE is the name of the extension point
+      // for post-invocation middleware
+      finished = await this.invokeMiddleware(context, {
+        extensionPoint: POST_INVOCATION_MIDDLEWARE,
+      });
+      if (finished) return;
+      this.send(response, result);
+    } catch (error) {
+      this.reject(context, error);
+    }
+  }
+}
+```
+
+It's also possible to inject multiple instances of `InvokeMiddleware`, each of
+which has its own extension point name.
+
+{% include code-caption.html content="src/sequence.ts" %}
+
+```ts
+export class SequenceWithMiddleware extends DefaultSequence {
+  /**
+   * Optional middleware chain
+   * Invokes registered middleware (injected via SequenceActions.MIDDLEWARE).
+   */
+  @inject('middleware.postInvoke', {optional: true})
+  protected middlewarePostInvoke: InvokeMiddleware = () => {};
+
+  async handle(context: RequestContext): Promise<void> {
+    try {
+      const {request, response} = context;
+      // The default middleware chain
+      await this.invokeMiddleware(context);
+      const route = this.findRoute(request);
+      const args = await this.parseParams(request, route);
+      const result = await this.invoke(route, args);
+
+      // The second middleware chain for post-invocation processing
+      context.bind('invocation.result').to(result);
```

Comment:
What does this do? Is it needed?

Reply:
This shows the possibility to access results from other actions inside a middleware.

Reply:
Added more comments.